processing/src/main/java/org/apache/druid/segment/data/FrontCodedIndexed.java

```diff
@@ -0,0 +1,393 @@
+  private final int lastBucketNumValues;
+  private final Comparator<ByteBuffer> comparator;
+
+  private FrontCodedIndexed(
+      ByteBuffer buffer,
+      Comparator<ByteBuffer> comparator,
+      int bucketSize,
+      int numBuckets,
+      int lastBucketNumValues,
+      boolean hasNull,
+      int adjustedNumValues,
+      int adjustIndex,
+      int div,
+      int rem,
+      int offsetsPosition,
+      int bucketsPosition
+  )
+  {
+    if (Integer.bitCount(bucketSize) != 1) {
+      throw new ISE("bucketSize must be a power of two but was[%,d]", bucketSize);
+    }
+    this.buffer = buffer;
+    this.comparator = comparator;
+    this.bucketSize = bucketSize;
+    this.hasNull = hasNull;
+    this.adjustedNumValues = adjustedNumValues;
+    this.adjustIndex = adjustIndex;
+    this.div = div;
+    this.rem = rem;
+    this.numBuckets = numBuckets;
+    this.offsetsPosition = offsetsPosition;
+    this.bucketsPosition = bucketsPosition;
+    this.lastBucketNumValues = lastBucketNumValues;
+  }
+
+  @Override
+  public int size()
+  {
+    return adjustedNumValues;
+  }
+
+  @Nullable
+  @Override
+  public ByteBuffer get(int index)
+  {
+    final int adjustedIndex;
+    // due to vbyte encoding, the null value is not actually stored in the bucket (no negative values), so we adjust
+    // the index
+    if (hasNull) {
+      if (index == 0) {
+        return null;
+      } else {
+        adjustedIndex = index - 1;
+      }
+    } else {
+      adjustedIndex = index;
+    }
+    // find the bucket which contains the value with maths
+    final int offsetNum = adjustedIndex >> div;
+    final int bucketIndex = adjustedIndex & rem;
+    // get offset of that bucket in the value buffer
+    final int offset = offsetNum > 0 ? buffer.getInt(offsetsPosition + ((offsetNum - 1) * Integer.BYTES)) : 0;
+    ByteBuffer copy = buffer.asReadOnlyBuffer().order(buffer.order());
+    copy.position(bucketsPosition + offset);
+    return getFromBucket(copy, bucketIndex);
+  }
+
+  @Override
+  public int indexOf(@Nullable ByteBuffer value)
+  {
+    // performs binary search using the first values of each bucket to locate the appropriate bucket, and then does
+    // a linear scan to find the value within the bucket
+    ByteBuffer copy = buffer.asReadOnlyBuffer().order(buffer.order());
+    if (value == null) {
+      return hasNull ? 0 : -1;
+    }
+    int minBucketIndex = 0;
+    int maxBucketIndex = numBuckets - 1;
+    while (minBucketIndex < maxBucketIndex) {
+      int currBucketIndex = (minBucketIndex + maxBucketIndex) >>> 1;
+      int currBucketFirstValueIndex = currBucketIndex * bucketSize;
+
+      final ByteBuffer currBucketFirstValue = get(currBucketFirstValueIndex + adjustIndex);
+      // we compare against the adjacent bucket to determine if the value is actually in this bucket or if we need
+      // to keep searching buckets
+      final ByteBuffer nextBucketFirstValue = get(currBucketFirstValueIndex + bucketSize + adjustIndex);
```

Comment:
Delegating to `get()` like this is going to create a bunch of extra objects as it will read into a `byte[]` and then wrap that in a `ByteBuffer` in order to return it.  You are just comparing things, so should be able to compare bytes directly using the `copy` that you already created without needing to create any new objects.  Should re-write this to not be based on get, but to push the comparison down to the `copy` buffer instead.

Reply:
I think i can do this if i assume the comparator, which I guess since we are currently only using this for strings would be ok to do. I was trying to make this somewhat generic, and I believe the buffer comparators are based on buffer position, so really would want to dissect the comparator and just do its stuff by hand so i can do it on the buffer in place rather than making the comparator be configurable.